### PR TITLE
encoding: refactor the character encoding 

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Related to #26814.

we declare some encoding variables like `utf8`, `gbk` etc. so that we can use these variables directly and easy to add a new one

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: parser/charset/encoding.go [hunk 1]
```diff
@@ -54,43 +52,34 @@ type Encoding struct {
...
-	return e.enc != nil && e.charLength != nil
+	return e != UTF8Encoding
 }
 
 // Name returns the name of the current encoding.
 func (e *Encoding) Name() string {
 	return e.name
 }
 
+// CharLength returns the name of the current encoding.
```
Comment (tangenta) [inline parser/charset/encoding.go:63]:
```suggestion
// CharLength returns next chararacter length in bytes.
```

File: parser/charset/encoding.go [hunk 2]
```diff
@@ -54,43 +52,34 @@ type Encoding struct {
...
-	if enc != nil && name != encodingLegacy {
-		e.enc = enc
-		e.charLength = FindNextCharacterLength(name)
-	} else {
-		e.enc = nil
-		e.charLength = nil
-	}
-	e.specialCase = LookupSpecialCase(e.name)
+func UpdateEncoding(label string) *Encoding {
+	return NewEncoding(label)
```
Comment (tangenta) [inline parser/charset/encoding.go:82]: Please remove this function and use NewEncoding instead.

Comment (xiongjiwei) [reply parser/charset/encoding.go]: updated

**Review by tangenta — APPROVED** [review]

**Review by zimulala — COMMENTED** [review]
Please briefly introduce the content of this PR process

File: parser/charset/gbk.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package charset
+
+import "golang.org/x/text/encoding/simplifiedchinese"
+
+var GBKEncoding = &Encoding{
+	enc:  simplifiedchinese.GBK,
+	name: "gbk",
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply parser/charset/gbk.go]: updated

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**